### PR TITLE
#164510298 Resolve database association with support for model autoloading

### PR DESCRIPTION
**Pull request by devdbrandy**
#### What does this PR do?

This PR resolves sequelize association misbehaviour and unhandled promises leading to test failure.

#### Description of Task to be completed?

- [x] Adjust test case for user authentication to return a promise
- [x] Autoload database models schemas
- [x] Stub mailer calls to external service
- [x] Write test against `validateConfigVariable` helper utils.

#### How should this be manually tested?

1. Run the following commands on terminal:

```bash
git clone -b feature/164510298/db-association-and-test-failure --single-branch \
https://github.com/andela/apollo-ah-backend.git

cd apollo-ah-backend
cp .env.example .env

# To run test
cp .env.example .env.test

# Install package modules
npm install
```

2. Modify `.env` file with your db credentials

3. Run database migrations with `npm run migrate`

4. Two npm scripts are available to run the test suite:

- `npm run mocha` or `npm run mocha:watch` - The later watches for any file changes and runs the full test suite (without code coverage)
- `npm test` - Performs a single full test suite run, including Istanbul code coverage reporting. Summary coverage reports are written to stdout, and detailed HTML reports are available in /coverage/index.html

**NOTE:** To run test locally, make sure to create `.env.test` file.

#### Any background context you want to provide?

Fix is based of the current changes on `staging` branch.

#### What are the relevant pivotal tracker stories?

[#164510298](https://www.pivotaltracker.com/story/show/164510298)

#### Screenshots

Screen Shot 2019-03-09 at 12 32 56 AM

#### Questions:

n/a

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -67,15 +67,16 @@ app.use('/api/v1', apiRoutes);
...
+  return res.send(swaggerSpec);
 });
 
 
 // / catch 404 and forward to error handler
 app.use((req, res, next) => {
   const err = new Error('Not Found');
   err.status = 404;
-  next(err);
+  console.log('oops', err)
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply index.js]: Removed.

File: controllers/usersController.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      if (user && !User.comparePassword(user, password)) {
+        throw createError(401, 'Invalid credentials');
+      }
+
+      // generate token from user payload
+      const payload = JSON.stringify(user.dataValues);
+      const token = await generateToken(payload);
+
+      // respond with token
+      return response
```
Comment (valenstical) [inline controllers/usersController.js]: **Nitpick**

- Use the `Response` helper function or follow the same output format to send responses.
This will ensure our responses are always in the same format.

Comment (devdbrandy) [reply controllers/usersController.js]: Oh, thought you must have implemented that. Nonetheless, I will make quick adjustments.

Comment (devdbrandy) [reply controllers/usersController.js]: Handle response using provided `responseHandler` helper function.

**Comment by valenstical** [comment]
Hello @devdbrandy, can you consider fixing the failing tests? You are doing a great job.

File: controllers/usersController.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
+import createError from 'http-errors';
+import jwt from 'jsonwebtoken';
+import models from '../models';
+import { env, generateToken } from '../helpers/utils';
+import Mail from '../helpers/sendMail';
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply controllers/usersController.js]: command `Mail.sendMail` was initially commented.

File: controllers/usersController.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    try {
+      const user = await User.create(body);
+      const token = await generateToken({ user });
+      // generate confirm token
+      const confirmToken = await generateToken({ email: user.email });
+      // generate confirm link
+      const confrimLink = `${env('API_DOMAIN')}api/v1/users/confirm_account?token=${confirmToken}`;
+      // send the user a mail
+      const data = {
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply controllers/usersController.js]: Resolved.

File: models/User.js [hunk 5]
```diff
@@ -47,8 +47,8 @@ const User = (sequelize, DataTypes) => {
    * @param {Sequelize.Model} models - Sequelize model
    * @returns {void}
    */
-  UserSchema.associate = (models) => {
-    // UserSchema.hasOne(models.Profile);
+  User.associate = (models) => {
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply models/User.js]: Added /* eslint-disable no-unused-vars */ at the top of file.

File: models/User.js [hunk 6]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable no-unused-vars */
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply models/User.js]: HoundeCI related.

File: test/integrations/routes/users.test.js [hunk 7]
```diff
@@ -21,27 +24,33 @@ describe('API endpoint: /api/users', () => {
   beforeEach(() => {
```
Comment (aanchirinah) [inline test/integrations/routes/users.test.js]: I really need to understand why you have to call `beforeEach` method to drop table and not put that logic in `after`

Comment (fejiroofficial) [reply test/integrations/routes/users.test.js]: like this  after(() => models.Profile.destroy({ truncate: true }));

Comment (fejiroofficial) [reply test/integrations/routes/users.test.js]: Please remove the sync block too, we have a migration script already

Comment (devdbrandy) [reply test/integrations/routes/users.test.js]: Thanks for the feedback @aanchirinah

It doesn't actually drop the table, what it simply does is that it clears the data on the database table `models.User.destroy()`. One of the benefits of doing is that, it prepares the database table regardless of whatever test cases may be introduced later in the future.

And every test cases is not dependant on another, which ensures that all test will always run on a clean, empty table.

Comment (devdbrandy) [reply test/integrations/routes/users.test.js]: In regards to that, the previous comment prior to this PR states that `// drop tables`, that was a typo. However, it has been adjusted to pass the right message.

Comment (devdbrandy) [reply test/integrations/routes/users.test.js]: I hope I have been able to tailor this in the right fashion (^_^). Looking forward to your feedback response, thanks.

Comment (devdbrandy) [reply test/integrations/routes/users.test.js]: Oh yes @fejiroofficial, you are very correct, thanks for pointing that, running the migration script has covered that. I will make a quick adjust to that, thanks.

Comment (fejiroofficial) [reply test/integrations/routes/users.test.js]: Okay good one

File: test/integrations/routes/users.test.js [hunk 8]
```diff
@@ -2,46 +2,51 @@
...
-  before(() => {
-    // synchronize and create tables
-    models.sequelize.sync();
-  });
   beforeEach(() => {
     // reset the dummy user data
     dummyUser = { ...DUMMY_USER };
-    // drop tables
+    // clear database table
     Bluebird.all([
```
Comment (fejiroofficial) [inline test/integrations/routes/users.test.js]: this part should be after the test has finished. You can truncate. But I strongly advise against this. The profile table and article table depends on this. Hence it is not cool if the other tests run and cannot find the data in the user table anymore because you truncated it. I strongly recommend we remove the truncate from the users test file.

Comment (devdbrandy) [reply test/integrations/routes/users.test.js]: I understand this may be a concern considering that the profile data depends on the user. Moving forward I think its a good idea to retain the data. Changes have been made.

Comment (fejiroofficial) [reply test/integrations/routes/users.test.js]: Thank you.

File: test/integrations/routes/users.test.js (hunk 7 above)
Comment (devdbrandy) [reply test/integrations/routes/users.test.js]: Hi @aanchirinah I have attended to your changes.

Comment (aanchirinah) [reply test/integrations/routes/users.test.js]: @devdbrandy Why have you removed both hooks?

**Comment by devdbrandy** [comment]
Will incorporate changes to you a new PR.
